### PR TITLE
Implement keepUnused config property

### LINKED ISSUE #12: Prevent removal of "unused" twin.macro
**Is your feature request related to a problem? Please describe.**
The [twin.macro](https://github.com/ben-rogerson/twin.macro) is from a family of babel-plugin-macros for TailwindCSS framework. In practice, it might look like.

```js
import tw from 'twin.macro'

export default function() {
  return <div tw="my-2" />
}
```

Unfortunately, that import is being removed as unused, but macros are a rather special beast similar to React.

**Describe the solution you'd like**
It would be probably best to cover this in `groupRules`, some different flag that prevents the removal of that group. The `React/h` could be handled this way too so it doesn't need to be hardcoded.

**Describe alternatives you've considered**
I tried the following, but it has no effect, and import is removed.

```js
"tsImportSorter.configuration.groupRules": [
  {
    "flag": "all",
    "regex": "^twin.macro$"
  },
  "^react(-dom)?$",
  {},
  "^[@]",
  "^[.]"
]
```

### PR DESCRIPTION
**Pull request by FredyC**
@daidodo I have run into a bit of a pickle and I could use advice on how to proceed further.

Instead of solving it inside the `sortImports` I thought I would actually filter the `usage` based on `keepUnused` and then `ImportNode.removeUnusedNames` would be considering only those unused that don't match that configuration. This is what I started with...

https://github.com/daidodo/tsimportsorter/compare/master...FredyC:keep-unused?expand=1#diff-5161f4a1c904cc67c6e75c2f19a6f2c5R26-R39

The problem is that `usage.unusedNames` is just a list of names without relation to import node or path. I cannot filter it like that.

I can think of two _bad_ solutions

 * Don't filter beforehand and include whole unused logic inside `ImportNode.removeUnusedNames`
 * Change `usage.unusedNames` to Map where the key would be a reference to ImportNode

Fixes #12

### CODE REVIEW

**Comment by daidodo** [comment]
@FredyC, very fast work! I can see almost the shape!

Instead of hacking into `usage`, I'd suggest to extract and transform `keepUnused` in [sortImports](https://github.com/daidodo/tsimportsorter/blob/master/src/format/sort/index.ts#L20), then pass it as a 2nd parameter to [n.removeUnusedNames](https://github.com/daidodo/tsimportsorter/blob/master/src/format/sort/index.ts#L24).
Now inside of [removeUnusedNames](https://github.com/daidodo/tsimportsorter/blob/master/src/format/parser/ImportNode.ts#L92), you have enough info to decide final unused names.

One edge case is that when multiple elements in `keepUnused` match the path, you should *combine* the names, e.g. `['abc']` + `[]` => `[]`.

**Comment by FredyC** [comment]
> One edge case is that when multiple elements in `keepUnused` match the path, you should _combine_ the names, e.g. `['abc']` + `[]` => `[]`.

I don't follow. You mean with a config like this it shouldn't match any names? How does that make sense? :)

```js
keepUnusued: [
  { path: "A", names: ["B"] },
  { path: "A", names: [] },  
]
```

Or is that a typo and you mean it will be `["B"]`. To be clear what about this situation?

```js
keepUnusued: [
  { path: "A", names: ["B"] },
  { path: "A", names: ["C"] },  
]
```

As a user, I would have expected to keep both `B` and `C` for matching path `A`.

**Comment by daidodo** [comment]
From the [specification](https://github.com/daidodo/tsimportsorter/issues/12#issuecomment-629577755):
> You need to match both paths and names to keep an unused name.
> If names is null or empty, all names are kept.

So that `[{"path": "abc", "names": ["Hi"]}]` keeps `Hi` from `abc`.
And `[{"path": "abc", "names": []}]` keeps all names from `abc`.

When *combining* the names:
* `["A"]` + `["B"]` ==> `["A", "B"]`
* `["A"]` + `[]` ==> `[]` (which means *all names*)

**Comment by FredyC** [comment]
Alright, I have a first working version now with passing tests. Have a look at it please and even you have some more test scenarios in mind, let me know. I will push docs update later.

FYI, about tests, I am running on Windows and some tests are failing because of CRLF vs LF, for example, `insert-lines/head/use-strict/keep`, the `0.origin.ts` is saved with CRLF, but `result.ts` with LF. I guess there is some intention to normalize it but doesn't work on Windows because CRLF is default. 🤷‍♂️

**Comment by FredyC** [comment]
> When _combining_ the names:



> From the [specification](https://github.com/daidodo/tsimportsorter/issues/12#issuecomment-629577755):
> 
> > You need to match both paths and names to keep an unused name.
> > If names is null or empty, all names are kept.
> 
> So that `[{"path": "abc", "names": ["Hi"]}]` keeps `Hi` from `abc`.
> And `[{"path": "abc", "names": []}]` keeps all names from `abc`.
> 
> When _combining_ the names:
> 
> * `["A"]` + `["B"]` ==> `["A", "B"]`
> * `["A"]` + `[]` ==> `[]` (which means _all names_)

I don't particularly agree with this, specifying the `[]` for names feels too magical in my opinion. If you want to keep all names, either omit `names` completely or set it to null. That's how it's done now. If the array is empty, it won't keep any names except default import if present.

**Comment by FredyC** [comment]
Hm, I actually tried to add a test case for the empty names array and behavior was kinda unexpected - it removed a whole node :)

So I've modified it the way you are proposing because it's easier. I personally don't care that much about that edge case to complicate the code for it more :)

File: src/format/config/index.ts [hunk 1]
```diff
@@ -23,6 +23,7 @@ export type Configuration = Readonly<
     maxDefaultAndBindingNamesPerLine: number;
     maxExportNamesPerLine: number;
     maxNamesPerWrappedLine: number;
+    keepUnused: KeepUnusedConfig[];
```
Comment (daidodo) [inline src/format/config/index.ts:26]: Please also update `package.json` and schemas (`schemas/`).

File: src/format/config/index.ts [hunk 2]
```diff
@@ -38,6 +39,13 @@ export type Configuration = Readonly<
   }>
 >;
 
+export type KeepUnusedConfig =
```
Comment (daidodo) [inline src/format/config/index.ts:42]: Nit-pick: The type name sounds a bit meaningless. Could you find a better name? Or what about `KeepRule`?

Comment (FredyC) [reply src/format/config/index.ts:42]: Well, I saw `ComposeConfig` just below it so it kinda inspired me :) So `KeepUnusedRule` then?

**Comment by FredyC** [comment]
Alright, updated schema, package, and readme - hopefully, it's enough like that. It's pretty much ready for the roll-out.

File: src/format/config/index.ts (hunk 1 above)
Comment (FredyC) [reply src/format/config/index.ts:26]: I was wondering where the VSCode is getting what settings are available. Hopefully, it's good enough like I did it now.

File: src/format/parser/KeepUnused.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import { KeepUnusedConfig } from '../config';
+import ImportNode from './ImportNode';
+
+export default class KeepUnused {
+  private readonly entries: Map<ImportNode, KeepUnusedMatcher[]>;
```
Comment (daidodo) [inline src/format/parser/KeepUnused.ts:5]: I don't think this map is needed and it introduces a dependency cycle with `ImportNode`.
An array of `KeepUnusedMatcher` should be good here. 
And this class needs to expose only one function:
```ts
/**
 * Returns a function to determine if a name is kept for the path.
 * Or undefined if no names will be kept.
 */
forPath(path: string): undefined | (name:string) => boolean;
```

File: README.md [hunk 4]
```diff
@@ -446,6 +452,25 @@ _Note:_
 
 For more details and how to construct your own rules, please read [the Wiki](https://github.com/daidodo/tsimportsorter/wiki/Sorting-Rules).
 
+### Unused imports removal
+
+By default all unused imports are removed. In some cases you might want to keep the import even if it's unused. For example to keep `import tw from 'twin.macro'` around you can do following:
+
+```json
+"keepUnused": ["twin.macro"]
+]
```
Comment (daidodo) [inline README.md:461]: Typo.

File: README.md [hunk 5]
```diff
@@ -446,6 +452,25 @@ _Note:_
 
 For more details and how to construct your own rules, please read [the Wiki](https://github.com/daidodo/tsimportsorter/wiki/Sorting-Rules).
 
+### Unused imports removal
+
+By default all unused imports are removed. In some cases you might want to keep the import even if it's unused. For example to keep `import tw from 'twin.macro'` around you can do following:
```
Comment (daidodo) [inline README.md:457]: following => the following

File: schemas/import-sorter.schema.json [hunk 6]
```diff
@@ -40,6 +40,12 @@
         "names": ["_", "aA"]
       }
     },
+    "keepUnused": {
+      "type": "array",
+      "default": [],
+      "items": { "type": ["string", "object"] },
```
Comment (daidodo) [inline schemas/import-sorter.schema.json:46]: This may not enough to provide hint for users. Plese be precise and detailed abouth what's in the `object`.

File: src/format/config/index.ts (hunk 1 above)
Comment (daidodo) [reply src/format/config/index.ts:26]: There is a `configuration` entry in `package.json` and that's where VS Code finds all settings about the extension.

File: src/format/config/index.ts (hunk 2 above)
Comment (daidodo) [reply src/format/config/index.ts:42]: I'm ok with KeepUnusedRule.

File: src/format/parser/ImportNode.ts [hunk 7]
```diff
@@ -262,14 +268,18 @@ export default class ImportNode extends Statement {
...
+  keepUnused: (name: string) => boolean,
+) {
   if (!name) return false;
   const { unusedNames, usedNames } = usage;
   const n = typeof name === 'string' ? name : name.aliasName ?? name.propertyName;
   if (!n) return false;
   // `unusedNames` (from TS compiler) gives more accurate results
   // than `usedNames` (from manual parsing).
-  if (unusedNames) return !unusedNames.has(n);
+  if (unusedNames) return !unusedNames.has(n) || keepUnused(n);
```
Comment (daidodo) [inline src/format/parser/ImportNode.ts:282]: `keepUnused(n)` should  be standalone and before this `if`.
```suggestion
  if (keepUnused(n)) return true;
  if (unusedNames) return !unusedNames.has(n);
```

File: src/format/parser/KeepUnused.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
+import { KeepUnusedConfig } from '../config';
+import ImportNode from './ImportNode';
+
+export default class KeepUnused {
+  private readonly entries: Map<ImportNode, KeepUnusedMatcher[]>;
+
+  constructor(cfg: KeepUnusedConfig[] = [], nodes: ImportNode[]) {
+    const matchers = cfg.map(entry => new KeepUnusedMatcher(entry));
```
Comment (daidodo) [inline src/format/parser/KeepUnused.ts:8]: Say you have entries defined as this:
```ts
private readonly entries_: { path: RegExp, names?: RegExp[] }[];
```
Then the following is all you want:
```suggestion
    this.entries_= cfg
      .filter(e => e.path)
      .map(e => ({path: new RegExp(e.path), names: e.names?.map(n => new RegExp(n))}));
```
And you can remove the `KeepUnusedMatcher` class, or simplify it as an interface.

File: src/format/config/index.ts (hunk 2 above)
Comment (daidodo) [inline src/format/config/index.ts:42]: Also I have a doubt whether this type needs to be exported.

File: src/format/sort/index.ts [hunk 9]
```diff
@@ -20,8 +21,9 @@ export function sortImports(
   const { sortRules: sort, groupRules: subGroups } = config;
   // The top group must be a match-all group.
   const group = new SortGroup({ flag: 'all', regex: '', sort, subGroups }, sorter);
+  const keepUnusedBouncer = new KeepUnused(config.keepUnused, nodes);
```
Comment (daidodo) [inline src/format/sort/index.ts:24]: Nit-pick: Please extract `keepUnused` from line 21 to keep consistency.

File: src/test/suite/examples/unused/import-sorter.json [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+    { "path": null },
+    { "path": "" },
+    { "path": "keepB" },
+    { "path": "keepC", "names": null },
+    { "path": "keepD", "names": ["keepD"] },
+    { "path": "keepE", "names": ["keepE"] },
+    { "path": "keepF", "names": [] },
+    { "path": "keepG", "names": ["keepG"] },
+    { "path": "keepH", "names": ["keepH"] }
+  ]
```
Comment (daidodo) [inline src/test/suite/examples/unused/import-sorter.json:14]: Lack of cases for multiple entries for a path, with different names.

(You don't need to keep them all in one `import-sorter.json`.)

File: src/test/suite/examples/unused/keep.result.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import A from 'keepA';
+import keepB, { B } from 'keepB';
+import * as any from 'keepC';
+import { keepD } from 'keepD';
+import { keepE } from 'keepE';
+import F1, { F2 } from 'keepF';
```
Comment (daidodo) [inline src/test/suite/examples/unused/keep.result.ts:6]: So you are keeping all names for `"names": []`?
I'm totally fine with it (actually lean towards it) but just want to confirm with you.

Comment (daidodo) [reply src/test/suite/examples/unused/keep.result.ts:6]: Just saw your reply and resolve this.

File: src/format/parser/KeepUnused.ts (hunk 3 above)
Comment (FredyC) [reply src/format/parser/KeepUnused.ts:5]: Well, it is needed because I need to pair ImportNode with matching lines from the configuration - basically to fulfill your use case above of merging multiple `path` that match single import. Feel free to experiment if you think it's not needed, commits are allowed from maintainers for the PR.

File: src/format/parser/KeepUnused.ts (hunk 8 above)
Comment (FredyC) [reply src/format/parser/KeepUnused.ts:8]: I don't think so, see my comment above why `Map` is necessary. I had a similar approach before and it did not work.

File: src/format/config/index.ts (hunk 2 above)
Comment (FredyC) [reply src/format/config/index.ts:42]: It's used in `KeepUnusedMatcher` constructor...

File: src/format/parser/ImportNode.ts (hunk 7 above)
Comment (FredyC) [reply src/format/parser/ImportNode.ts:282]: I guess if it's supposed to be kept it doesn't matter if it's unused or not, good call.

File: schemas/import-sorter.schema.json (hunk 6 above)
Comment (FredyC) [reply schemas/import-sorter.schema.json:46]: I wasn't able to find any documentation on this, so I mostly picked from your other pieces. However, something is still not right. When I run it and try to edit config, it looks like this.

image

File: src/test/suite/examples/unused/import-sorter.json (hunk 10 above)
Comment (FredyC) [reply src/test/suite/examples/unused/import-sorter.json]: I don't follow what case you have in mind?

Comment (FredyC) [reply src/test/suite/examples/unused/import-sorter.json]: I've added another case, hopefully that's what you meant?

File: schemas/import-sorter.schema.json [hunk 12]
```diff
@@ -156,6 +162,22 @@
...
+    },
+    "keepUnusedRule": {
+      "oneOf": [
+        { "type": "string" },
+        {
+          "type": "object",
+          "properties": {
+            "path": { "type": "string", "description": "Import path pattern to keep" },
+            "names": {
+              "type": ["array", "null"],
```
Comment (daidodo) [inline schemas/import-sorter.schema.json:174]: Nit: Is `null` needed here?

File: src/test/suite/extension.test.ts [hunk 13]
```diff
@@ -38,7 +38,7 @@ suite('Integration Test Suite', () => {
   const examples = getTestSuite(dir, 'examples');
   if (!examples) return;
   // Run all tests
-  return runTestSuite(examples);
+  return runTestSuite(examples, 'unused/keep');
```
Comment (daidodo) [inline src/test/suite/extension.test.ts:41]: Please don't commit this file.

Comment (FredyC) [reply src/test/suite/extension.test.ts]: Whoops, sorry :) Wish there was a better way than editing that file. Perhaps another file that is git ignored and value is read from it if it's present?

File: src/test/suite/examples/unused/keep/import-sorter.json [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+    { "path": "keepB" },
+    { "path": "keepC", "names": null },
+    { "path": "keepD", "names": ["keepD"] },
+    { "path": "keepE", "names": ["keepE"] },
+    { "path": "keepF", "names": [] },
+    { "path": "keepG", "names": ["keepG"] },
+    { "path": "keepH", "names": ["keepH"] },
+    { "path": "keepI", "names": ["keepI1"] },
+    { "path": "keepI", "names": ["keepI2"] }
+  ]
```
Comment (daidodo) [inline src/test/suite/examples/unused/keep/import-sorter.json:16]: More cases when `names` is `null` or empty, as well as in different orders.

File: src/format/parser/KeepUnused.ts (hunk 3 above)
Comment (FredyC) [reply src/format/parser/KeepUnused.ts:5]: Also, to be honest, after spending nearly half a day on this, I don't feel like rewriting it for some uncertain gain. Given how rarely this whole thing will be used (since React/h is handled by default), I don't it's worth spending more time on it without a good reason.

File: src/test/suite/examples/unused/keep/import-sorter.json (hunk 14 above)
Comment (FredyC) [reply src/test/suite/examples/unused/keep/import-sorter.json:16]: I don't know, it feels like this covers most major and some minor cases. I don't really think people will try to get too wild here and trying to break it :) Most of them will use string names for whole modules probably, lol :)

**Comment by FredyC** [comment]
As I said [above](https://github.com/daidodo/tsimportsorter/pull/13#discussion_r426280175), I don't feel like spending time on this already. I consider the current state good enough given the rare cases of when this will be used.

Of course, feel free to dig in and rewrite a whole thing if you like. I tried my best and I am satisfied with it 💪

**Comment by FredyC** [comment]
Wow, you really did rewrite it and it still passes all tests. Amazing how you managed to simplify it. I guess I was bit overengineering it :)

**Comment by FredyC** [comment]
Although, readability is arguable, with all those one letter variables I am totally lost in there, but it's your choice if you like it that way.

**Comment by daidodo** [comment]
I changed a few code and added more test cases. Now I'm ok to merge it.

**Comment by daidodo** [comment]
Thank you very much for the contribution!

**Comment by FredyC** [comment]
And I thank you for finishing it up. Hopefully, you can push a new version soon so I can get rid of those disable comments :)
